[2/N] support of replication fallback strategy

zpcore opened the pull request:
Support implicit replication strategy registration.

To enable implicit strategy registration, call `implicit_strategy_context()` as below:
1. Use default replication strategy for all missing strategy ops:
```
with implicit_strategy_context():
    ...
```
2. Specify custom strategy as input for centain ops. Those strategies will be registered at runtime. Otherwise, default replication strategy will apply:
```
with implicit_strategy_context(
            [ OpStrategyPack(custom_op_1.default, custom_strategy_1, schema_info=None),
              OpStrategyPack(custom_op_2.default, custom_strategy_2, schema_info=None)
              ...]
        ):
    ...
```

Note for AutoParallel: I also split out the `try_propagate_implicit_strategy()` to help automatic register strategy if missing. We should replace all `torch.distributed.tensor.DTensor._op_dispatcher.sharding_propagator.op_strategy_funcs[op](op_schema)` with `torch.distributed.tensor.DTensor._op_dispatcher.sharding_propagator.try_propagate_implicit_strategy(op, op_schema)` to enable the feature. 

PR stack:
[2/N] support of replication fallback strategy https://github.com/pytorch/pytorch/pull/158476
[1/N] support of replication fallback strategy https://github.com/pytorch/pytorch/pull/158046

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci @fmassa

zpcore committed:
(1/N) support of replication fallback strategy

fix per fb

create a test function to check if strategy output identical strategy

nit

use expand_to_full_mesh_op_strategy to build replicate strategy

lint

fix flatten function

zpcore committed:
Support implicit strategy registration

wconstab commented:
nit on naming:  implicit_strategy_context is not clear to users

you can drop the context part bc it is clear from usage that it is a ctxmgr when you see someone use it as "with _"

i'd like the name to convey something about
 - replication
 - unsupported operators

maybe `with replicate_for_unsupported_operators()`?  maybe others will have a better suggestion

wconstab commented on torch/distributed/tensor/_ops/utils.py:
```diff
@@ -381,3 +382,58 @@ def expand_to_full_mesh_op_strategy(
+
+    op_overload: torch._ops.OpOverload
+    strategy_function: Callable[[OpSchema], StrategyType]
+    schema_info: Optional[RuntimeSchemaInfo] = None
```
hm, im wondering if we need to introduce a new class here.

1) should we get rid of 'strategy_function' and have it always be replicate?  (there is another way to register a strategy and perhaps we don't want to 'abuse' the new implicit ctx for things besides replication strategy?

2) what is schema_info needed for?  it wasn't used in the tests above, and naively i was thinking of just registering it by op_overload alone

zpcore replied:
1. I am thinking that once we have a new universal strategy (e.g., `shard_batch_strategy`) to support shard on specific batch, then they can register the `shard_batch_strategy` here instead of falling back to replicate strategy.

2. We may need to check `schema_info.needs_pytree` to flatten TupleStrategy.

wconstab commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -254,6 +259,32 @@ def spec_to_strategy(spec: object) -> object:
             kwargs_schema=kwargs_op_strategy,
         )
 
+    def try_propagate_implicit_strategy(
```
nit: this function is essentially "the strategy" function in that it creates an op_strategy.

by convention, other ops name their strategy functions like this: `def topk_strategy(op_schema: OpSchema) -> OpStrategy:`.  

Should we therefore just name this function 'generic_replicate_strategy'?

zpcore replied:
I split out the code `try_propagate_implicit_strategy` so that AP can call. If AP use `torch.distributed.tensor.DTensor._op_dispatcher.sharding_propagator.op_strategy_funcs[op](op_schema)` as it is now, we won't be able to register any implicit strategy for them. In general, what AP needs is:

-------
with implicit_strategy_context():
 &nbsp; &nbsp; &nbsp; &nbsp; ...
    &nbsp; &nbsp; &nbsp; &nbsp; ~`torch.distributed.tensor.DTensor._op_dispatcher.sharding_propagator.op_strategy_funcs[op](op_schema)`~ &rarr;` implicit_strategy_context(op, op_schema)`
 &nbsp; &nbsp; &nbsp; &nbsp; ...

wconstab replied:
Oh, I think we can just directly use replicate_strategy function in AP too.

zpcore replied:
I think `try_propagate_implicit_strategy` above is the API we want as requested here: https://github.com/pytorch-labs/autoparallel/pull/29#discussion_r2196205840. AP for sure can use `replicate_strategy` directly, but they need to specify ops and that is not implicit call right?

wconstab commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -278,10 +309,56 @@ def propagate_op_sharding_non_cached(self, op_schema: OpSchema) -> OutputShardin
             strategy_schema = self._wrap_with_op_strategy(op_schema)
 
+            # assign implicit strategy if enabled
+            self.try_propagate_implicit_strategy(op_schema.op, strategy_schema)
```
hmm. if ` not self.enable_implicit_strategy` is true and we call this, won't we raise an exception?

would a clearer way to do this be

```
if op_schema.op in self.op_strategy_funcs:
    op_strategy = self.op_strategy_funcs[op_schema.op](strategy_schema)
else:
   op_strategy = replicate_op_strategy
```

zpcore replied:
> if  not self.enable_implicit_strategy is true and we call this, won't we raise an exception?

This is expected since the original code will also raise the exception. I basically move the following `raise` into the `try_propagate_implicit_strategy `:
https://github.com/pytorch/pytorch/blob/e311886e3d57c83a88b97a084dd0b95d6d1537a8/torch/distributed/tensor/_sharding_prop.py#L531-L533

zpcore committed:
fix pytree for TupleStrategy

zpcore commented on test/distributed/tensor/test_op_strategy.py:
```diff
@@ -662,5 +664,40 @@ def test_tuple_replicate_strategy_placement(self):
+
+        # 4. runtime specify test_op strategy
+        # TODO(zpcore): try with a different universal strategy once we have
+        with implicit_strategy_context(
```
Looks like our understanding of the context usage is different. 

Suppose we have op: `op1`, `op2`, `ops` (all not registered) and strategies `s1`. My intention is that if we call
```
with implicit_strategy_context([ OpStrategyPack (op1, s1)]):
```
Then `op1` will use strategy `s1`; `op2` and `op3` will fall back to replication strategy. This is what the code is currently doing.

I think what you mean is that in this case `op1` will use strategy `s1` while `op2` and `op3` are still unregistered.

wconstab replied:
I think it's a little confusing to mix those in the same context manager call.

For the batch dim sharding it might be best to prototype it before tying it into the api here. I'm wondering how a user will specify which input(s) of the op and which dim of the input to try sharding. Maybe, we just create a separate context manager for the batch rule?

zpcore replied:
For batch dim sharding, my plan is to make a strategy to take `batch_dim` as `replicate_with_batch_dim_shard_strategy(batch_dim)(op, op_schema)`. I plan to get a prototype in the follow up PR. Then if we use a seperate context manger for the batch sharding rule, different op may require different `batch_dim`. We will probably end up with still needing to specify strategy seperately for ops as below:
```
with implicit_batch_dim_shard_strategy_context([ (op1, dim 0), (op2, dim 1),... ])
 ...
```

Maybe a better solution is to just let `implicit_strategy_context()` don't take any argument and default any missing ops to replication strategy as you suggested. If we want to use `implicit_batch_dim_shard_strategy_context`, user must regisiter those op explicitly with batch dim specification. No need to create another context manager. Does this look reasonable?

wconstab replied:
My point was that it's not even as simple as (op1, dim1) because op1 may have 3 inputs and the batch dim might be (dim0, dim1, None) for the 3 inputs. Or were you only planning to support single input ops? Maybe there isn't actually an example of different batch dim on 2 inputs but there should be at least ops that have some inputs without any batch dim. 

I think keeping the context manager simpler for now is better. You can always add kwargs to it later or add a separate context manager if you want.

wanchaol reviewed (commented):
I feel these changes related to replicate fallback strategy are becoming too intrusive to DTensor.

I don't feel it's very useful to make this a feature of DTensor. It might be meaningful to the auto parallel project, but I would recommend putting these changes in the auto parallel repo directly instead of DTensor (i.e. you can put these changes to a dtensor_utils.py file).

In general I feel that in DTensor we can implement more operators to improve coverage and fix bugs to any existing features to unblock, but for things specific to auto parallel they should live in the auto parallel repo, that's where you can iterate faster too :)

zpcore commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -278,9 +309,56 @@ def propagate_op_sharding_non_cached(self, op_schema: OpSchema) -> OutputShardin
         out_tensor_meta = self._propagate_tensor_meta_non_cached(op_schema)
 
-        if op_schema.op in self.op_strategy_funcs:
+        if op_schema.op in self.op_to_rules:
```
Oh, actually this is just moving the `p_schema.op in self.op_to_rules` condition to the front. The real change is ~10 lines under the `else` condition.

zpcore commented on torch/distributed/tensor/_ops/utils.py:
```diff
@@ -332,3 +361,58 @@ def expand_to_full_mesh_op_strategy(
+
+
+@contextmanager
+def implicit_strategy_context(
```
We want to make this public (similar to the `def implicit_replication()` function) to make code mostly runnable with undefined op strategy.

wconstab replied:
I think for now it's ok to only offer implicit_replicate API. 

We also discussed offering a batch-shard API. The idea would be users mark the batch dims of inputs and those get sharded while others get replicated. The idea was from @fmassa. But we did not prototype this or design it yet so I think it's premature to offer a generalized contextmanager. If the batch API seems like a good idea we can directly try it and keep it separate from the implicit replication API.

wanchaol replied:
hmmm the batch-shard thing is sth I've already tried back in the days of dist compiler.. my conclusion is that it's not going to work well, that's why I think the replication fallback strategy is the only fallback strategy that is possible for DTensor.

zpcore replied:
Hi @wanchaol , can you elaborate more on why batch sharding strategy is "not going to work well"? Will it be easier if user can specify which dim of the arg input tensor is the batch dim? I think once user specify which dim is the batch dim, then the strategy can try shard the dim on different device axis, but it must make sure all those batch dim from different args must shard on the same device axis (or axises).

wanchaol replied:
@zpcore My conclusion is mainly: 1. the batch dim sharding is not quite reliable, unless you let user specify the batch dim for every or confused operations, as how to propagate the batch sharding is very hard to reason 2. even though in forward you can derive a way, the backward becomes inherently complicated than forward as you would have to think about partial placement.

zpcore commented:
> I feel these changes related to replicate fallback strategy are becoming too intrusive to DTensor.
> 
> I don't feel it's very useful to make this a feature of DTensor. It might be meaningful to the auto parallel project, but I would recommend putting these changes in the auto parallel repo directly instead of DTensor (i.e. you can put these changes to a dtensor_utils.py file).
> 
> In general I feel that in DTensor we can implement more operators to improve coverage and fix bugs to any existing features to unblock, but for things specific to auto parallel they should live in the auto parallel repo, that's where you can iterate faster too :)

Hmm, let's also keep @XilunWu and @wconstab in the discussion. Those feature requests are from AutoParallel that want us to upstream. I agree that we can keep those changes in AP if we want to keep DTensor as a basic building block for distributed tensor operations.

wanchaol commented:
@zpcore to clarify more why I don't think this is quite useful for DTensor, I think DTensor is already quite "magical", in the sense that system developers usually don't really know what communication happened under the hood, and if we always fall back to replication and make the code mostly runnable, it's going to create a whole blackbox for users where they would blame it's runnable but not very efficient, and don't know where's the inefficient sharding.

So I think we want DTensor to be as explicit as possible, it's not very good UX the we would have this replication fallback to happen by default if the op does not have a sharding strategy, it's better to tell user what's wrong. So I think it can happen only if user requested to have this kind of behavior, If we want to do this I would suggest merge this behavior with the `implicit_replication` API you mentioned

zpcore commented:
> @zpcore to clarify more why I don't think this is quite useful for DTensor, I think DTensor is already quite "magical", in the sense that system developers usually don't really know what communication happened under the hood, and if we always fall back to replication and make the code mostly runnable, it's going to create a whole blackbox for users where they would blame it's runnable but not very efficient, and don't know where's the inefficient sharding.
> 
> So I think we want DTensor to be as explicit as possible, it's not very good UX that we would have this replication fallback to happen by default if the op does not have a sharding strategy, it's better to tell user what's wrong. So I think it can happen only if user requested to have this kind of behavior, If we want to do this I would suggest merge this behavior with the `implicit_replication` API you mentioned

Thanks for the reply! Originally I thought the design philosophy is to hide low level details from the dtensor user. Once we consider "efficient > runnable", this completely makes sense to prevent such fallback in the background. I am addressing those review feedbacks and working on moving this PR to Autoparallel.
